src/librbd/api/Mirror.cc

```diff
@@ -3042,6 +3030,132 @@ int Mirror<I>::group_disable(IoCtx& group_ioctx, const char *group_name,
+
+  if (requires_orphan != nullptr) {
+    *requires_orphan = false;
+  }
+
+  auto snap =  group_snaps->rbegin();
+  for (; snap != group_snaps->rend(); ++snap) {
+    auto mirror_ns = std::get_if<cls::rbd::GroupSnapshotNamespaceMirror>(
+        &snap->snapshot_namespace);
+    if (mirror_ns == nullptr) {
+      continue;
+    }
+    ldout(cct, 20) << "previous snapshot snap_id=" << snap->id << " "
+                   << *mirror_ns << dendl;
+
+    if (mirror_ns->is_demoted() && !force) {
+      lderr(cct) << "trying to create primary snapshot without force "
+                 << "when previous primary snapshot is demoted"
+                 << dendl;
+      return false;
+    }
+
+    if (mirror_ns->state == cls::rbd::MIRROR_SNAPSHOT_STATE_NON_PRIMARY) {
+      if (!force) {
+        lderr(cct) << "trying to create primary snapshot without force "
+                   << "when previous snapshot is non-primary"
+                   << dendl;
+        return false;
+      }
+      if (demoted) {
+        lderr(cct) << "trying to create primary demoted snapshot "
+                   << "when previous snapshot is non-primary"
+                   << dendl;
+        return false;
+      }
+
+      if (requires_orphan != nullptr) {
+        *requires_orphan = !mirror_ns->is_demoted();
+      }
+      if (snap->state != cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE) {
+        ldout(cct, 20) << "needs rollback" << dendl;
+        if (!get_rollback_group_snap_id(++snap, group_snaps->rend(),
+                                        rollback_group_snap_id)) {
+          lderr(cct) << "cannot rollback" << dendl;
+          return false;
+        }
+        ldout(cct, 20) << "rollback_group_snap_id=" << rollback_group_snap_id << dendl;
+      }
+      return true;
+    }
+
+    return true;
+  }
+
+  ldout(cct, 20) << "no previous mirror snapshots found" << dendl;
+  return true;
+}
+
+int create_orphan_group_snapshot(IoCtx& group_ioctx,
+                                 std::string &group_id,
+                                 std::string &global_group_id) {
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  std::string group_snap_id = librbd::util::generate_image_id(group_ioctx);
+  ldout(cct, 20) << "io_ctx=" << &group_ioctx
+		 << ", group_id=" << group_id
+                 << ", group_snap_id=" << group_snap_id
+                 << dendl;
+
+  cls::rbd::GroupSnapshot group_snap{
+      group_snap_id,
+      cls::rbd::GroupSnapshotNamespaceMirror{
+        cls::rbd::MIRROR_SNAPSHOT_STATE_NON_PRIMARY, {}, {} , {}},
+      prepare_non_primary_mirror_snap_name(cct, global_group_id, group_snap_id),
+      cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE};
+
+  std::string group_header_oid = librbd::util::group_header_name(group_id);
+  int r = cls_client::group_snap_set(&group_ioctx,
+                                     group_header_oid, group_snap);
+  if (r < 0) {
+    lderr(cct) << "failed to set group snapshot metadata: " << cpp_strerror(r)
```

Comment:
The 2 err messages are identical - it will be difficult to figure out which step failed.
Perhaps "failed to create group snapshot" and "failed to mark snapshot complete"?